apps/rush-lib/src/cli/actions/AddAction.ts

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as os from 'os';
+
+import {
+  CommandLineFlagParameter,
+  CommandLineStringParameter
+} from '@microsoft/ts-command-line';
+
+import { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { BaseRushAction } from './BaseRushAction';
+import { RushCommandLineParser } from '../RushCommandLineParser';
+import { DependencyIntegrator, SemVerStyle } from '../../logic/DependencyIntegrator';
+
+export class AddAction extends BaseRushAction {
+  private _exactFlag: CommandLineFlagParameter;
+  private _caretFlag: CommandLineFlagParameter;
+  private _devDependencyFlag: CommandLineFlagParameter;
+  private _makeConsistentFlag: CommandLineFlagParameter;
+  private _noInstallFlag: CommandLineFlagParameter;
+  private _packageName: CommandLineStringParameter;
+  private _versionSpecifier: CommandLineStringParameter;
+
+  constructor(parser: RushCommandLineParser) {
+    const documentation: string[] = [
+      'Blah.'
+    ];
+    super({
+      actionName: 'add',
+      summary: 'Adds a dependency to the package.json and runs rush upgrade.',
+      documentation: documentation.join(os.EOL),
+      safeForSimultaneousRushProcesses: false,
+      parser
+    });
+  }
+
+  public onDefineParameters(): void {
+    this._packageName = this.defineStringParameter({
+      parameterLongName: '--package',
+      parameterShortName: '-p',
+      required: true,
+      argumentName: 'PACKAGE_NAME',
+      description: '(Required) The name of the package which should be added as a dependency'
+    });
+    this._versionSpecifier = this.defineStringParameter({
+      parameterLongName: '--version',
+      parameterShortName: '-v',
+      argumentName: 'VERSION_RANGE',
+      description: ''
+    });
```

Comment:
Can this be specified as `package@version`?

Reply:
Could be, but requires extra parsing and whatnot. We will make it like that when we allow anonymous arguments.

---
In reply to: [220407345](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220407345) [](ancestors = 220407345)

Reply:
Don't we already have parsing for that?

---
In reply to: [220743394](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220743394) [](ancestors = 220743394,220407345)

Reply:
Not that I am aware of...

---
In reply to: [220749976](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220749976) [](ancestors = 220749976,220743394,220407345)

Reply:
Can't we just split by the "@" character?

---
In reply to: [220751648](https://github.com/Microsoft/web-build-tools/pull/843#discussion_r220751648) [](ancestors = 220751648,220749976,220743394,220407345)